src/actions/BringMoveSwap.ts

```diff
@@ -0,0 +1,266 @@
+import {
+  Action,
+  ActionPreferences,
+  ActionReturnValue,
+  Graph,
+  TypedSelection,
+} from "../Types";
+import { runForEachEditor } from "../targetUtils";
+import update from "immutability-helper";
+import displayPendingEditDecorations from "../editDisplayUtils";
+import { computeChangedOffsets } from "../computeChangedOffsets";
+import { flatten, zip } from "lodash";
+import { Selection, TextEditorDecorationType, TextEditor } from "vscode";
+
+interface DecorationTypes {
+  sourceStyle: TextEditorDecorationType;
+  sourceLineStyle: TextEditorDecorationType;
+  destinationStyle: TextEditorDecorationType;
+  destinationLineStyle: TextEditorDecorationType;
+}
+
+interface Edit {
+  editor: TextEditor;
+  range: Selection;
+  newText: string;
+  targetsIndex: number;
+  originalSelection: TypedSelection;
+}
+
+interface ThatMarkEntry {
+  editor: TextEditor;
+  targetsIndex: number;
+  typedSelection: TypedSelection;
+  selection: Selection;
+}
+
+class BringMoveSwap implements Action {
+  targetPreferences: ActionPreferences[] = [
+    { insideOutsideType: "inside" },
+    { insideOutsideType: "inside" },
+  ];
+
+  constructor(private graph: Graph, private type: string) {
+    this.run = this.run.bind(this);
+  }
+
+  private broadcastSource(
+    sources: TypedSelection[],
+    destinations: TypedSelection[]
+  ) {
+    if (sources.length === 1 && this.type !== "swap") {
+      // If there is only one source target, expand it to same length as
+      // destination target
+      return Array(destinations.length).fill(sources[0]);
+    }
+    return sources;
+  }
+
+  private getDecorationStyles(): DecorationTypes {
+    if (this.type === "use") {
+      return {
+        sourceStyle: this.graph.editStyles.referenced,
+        sourceLineStyle: this.graph.editStyles.referencedLine,
+        destinationStyle: this.graph.editStyles.pendingModification0,
+        destinationLineStyle: this.graph.editStyles.pendingLineModification0,
+      };
+    }
+    return {
```

Comment:
For "move", I wonder if we want to use the `delete` style after?  Maybe that's what happens already

Reply:
Ahh no its modify now. I disnt even realize they had different colors ^^

Reply:
Btw after? There are no sources decoration after since the sources are gone. Nothing to decorate.

Reply:
yeah was thinking bout that.  I wonder if we should highlight it as a delete before then?

Reply:
Agreed